gossip/util/misc.go

```diff
@@ -187,6 +190,8 @@ func RandomInt(n int) int {
 // If we want a rand that's non-global and specific to gossip, we can
 // establish one. Otherwise this uses the process-global locking RNG.
 func RandomUInt64() uint64 {
+	rMutex.Lock()
+	defer rMutex.Unlock()
```

Comment:
1. Above r is used in 2 more places without lock. Why is lock not used there?

2. yes, this is a bug in the math/rand package. But after the release of v3.0.0 the package used https://github.com/hyperledger/fabric/commit/9b28d51052ce670bc70a244897e0712a4e0ab08b was changed, this change is not in the release yet. And your stacktrace is not using v2 code.

3. Need proof that you actually fixed the bug suddenly it was fixed by changing the package to version v2. You need a test to prove that your fixes are correct

```
/usr/local/go/src/math/rand/rng.go:249 +0x7b
math/rand.(*Rand).Uint64(0xc0180cf980?)
/usr/local/go/src/math/rand/rand.go:104 +0x22
```

Reply:
1. Above r is used in 2 more places without lock. Why is lock not used there?

2. yes, this is a bug in the math/rand package. But after the release of v3.0.0 the package used https://github.com/hyperledger/fabric/commit/9b28d51052ce670bc70a244897e0712a4e0ab08b was changed, this change is not in the release yet. And your stacktrace is not using v2 code.

3. Need proof that you actually fixed the bug suddenly it was fixed by changing the package to version v2. You need a test to prove that your fixes are correct

```
/usr/local/go/src/math/rand/rng.go:249 +0x7b
math/rand.(*Rand).Uint64(0xc0180cf980?)
/usr/local/go/src/math/rand/rand.go:104 +0x22
```

Reply:
> This is the panic that happens when there's a lot of concurrency:
> 
> panic: runtime error: index out of range [-1]
> 
> goroutine 3305 [running]: math/rand.(*rngSource).Uint64(0xc014386101?) /usr/local/go/src/math/rand/rng.go:249 +0x7b math/rand.(*Rand).Uint64(0xc0180cf980?) /usr/local/go/src/math/rand/rand.go:104 +0x22 github.com/hyperledger/fabric/gossip/util.RandomUInt64(...) /gossip/util/misc.go:187 github.com/hyperledger/fabric/gossip/gossip/algo.(*PullEngine).newNONCE(0xc006953000) /gossip/gossip/algo/pull.go:324 +0x25 github.com/hyperledger/fabric/gossip/gossip/algo.(*PullEngine).initiatePull(0xc006953000) /gossip/gossip/algo/pull.go:179 +0x11b github.com/hyperledger/fabric/gossip/gossip/algo.NewPullEngineWithFilter.func1() /gossip/gossip/algo/pull.go:126 +0x27 created by github.com/hyperledger/fabric/gossip/gossip/algo.NewPullEngineWithFilter in goroutine 1 /gossip/gossip/algo/pull.go:120 +0x2d5
> 
> I suppose the problem is because we are accessing this function a lot of times, so adding a lock ensures that there are no concurrency issues

If there is a panic stemming from gossip then I fail to see how is the chaincode related to the problem. 

We shouldn't fix two different problems in the same PR. 

Can you move all the non gossip related code changes to a new PR?

Reply:
yeah, but you see the problem, right? gossip is invoking randomUint64, and it panicking,

Reply:
you are right, these are two different problems for two different panic errors

Reply:
panic calls a different code than the one you're correcting.

Reply:
> So @pfi79 compiling the code in main should fix the issue right about gossip right?

It's possible. After updating the package to v2 I stopped catching the panic error in tests.